### DISCUSSION TITLE
NAV-RELPOSNED heading doesn't seem to be updating

### DISCUSSION THREAD

**Discussion by omuhr** [semuconsulting/pyubx2#121] [Q&A] [answered]
I have two ZED-F9P's set up in a moving base configuration, I am now trying to interface with the rover programatically in order to perform long term logging. Everything seems to work fine and I am able to parse the NAV-RELPOSNED messages and extract, among others, the relPosHeading field which I am after. But the problem is that the relPosHeading field doesn't seem to change whenever the antennae are rotated relative to each other. I know I am not reading the same message over and over, as the angle does change but only by a few decimal points when the platform is actually rotated by 15-30°. If I monitor the heading in u-center it does update in the expected way.

Here is a minimalist example of the code:

```python
import serial
import serial.tools.list_ports as list_ports
from pyubx2 import UBXReader

try:
    stream = serial.Serial('COM7', 9600, timeout=3)

except serial.serialutil.SerialException:
    ports = list_ports.comports()
    for port, desc, hwid in sorted(ports):
        print("{}: {} [{}]".format(port, desc, hwid))

ubr = UBXReader(stream)

for angle in list_of_angles:
    data = []
    rotate_platform() # This line would rotate the antennae relative to eachother
    while True:
        (raw_data, parsed_data) = ubr.read()
        if parsed_data.identity == 'NAV-RELPOSNED':
            data.append(parsed_data.relPosHeading)
            break
    log_data(data)
```

Any ideas on where I might be going wrong? Does the receiver need to be given any messages in order to properly update the NAV-RELPOSNED message?

Regards,

Oscar

**Comment by semuadmin** [comment]
Hi @omuhr 

Firstly, are you connecting to your F9P via UART or USB? If UART you might want to check the baud rate is correct - the default F9P UART baud rate is 38400, not 9600.

Your example code is fine as far as it goes, though one potential problem is your timeout value (_which may be causing you to miss messages arriving more quickly than every 3 seconds_) - see https://github.com/semuconsulting/pyubx2/blob/master/examples/ubxpoller.py for a more robust stream reader example.

Once you have the streaming mechanism sorted, you can simply read the parsed NAV-RELPOSNED message and reference any attributes you want - in your case `relPosHeading`. The raw heading value is scaled by 1e-5 degrees, but pyubx2 should apply this scaling automatically (assuming you haven't set the `scaling` keyword argument to `False`).

Could you provide the ACTUAL, COMPLETE Python script you are executing and an example of the output you're getting?

Could you also send me a screen shot of the comparative results you're getting in u-center, which you believe to be correct?

What values are you seeing for heading accuracy (e.g. `accHeading`)?

Thanks.

**Comment by semuadmin** [comment, marked as answer]
By way of illustration, I measured my own roving F9P against two different base locations in quick succession and compared the NAV-RELPOSNED `relPosHeading` and `relPosLength` readings with my own estimates of heading and distance (_derived using the `bearing()` and `haversine()` helper functions provided by pyubx2_).

```
"""
BASE1 - JoeSeelsGPS - 53.53, -1.33 60.9km

<UBX(NAV-RELPOSNED, version=1, reserved0=0, refStationID=0, iTOW=06:36:25, relPosN=-879166, relPosE=-6068417, relPosD=29273, relPosLength=6131841, relPosHeading=261.7566, reserved1=0, relPosHPN=0.0, relPosHPE=-3.9, relPosHPD=2.4, relPosHPLength=3.9, accN=270.1, accE=292.4, accD=493.1, accLength=291.9, accHeading=0.00025, reserved2=0, gnssFixOK=1, diffSoln=1, relPosValid=1, carrSoln=1, isMoving=0, refPosMiss=0, refObsMiss=0, relPosHeadingValid=1, relPosNormalized=0)>

reported relPosLength = 6131841.039
reported relPosHeading = 261.7566

BASE2 - ER_Pollybell_1 - 53.43, -0.91 88.2km

<UBX(NAV-RELPOSNED, version=1, reserved0=0, refStationID=2690, iTOW=06:37:30, relPosN=267987, relPosE=-8849794, relPosD=58193, relPosLength=8854041, relPosHeading=271.7345, reserved1=0, relPosHPN=5.1, relPosHPE=0.0, relPosHPD=9.0, relPosHPLength=7.8, accN=209.3, accE=217.4, accD=298.3, accLength=217.4, accHeading=0.00014, reserved2=0, gnssFixOK=1, diffSoln=1, relPosValid=1, carrSoln=1, isMoving=0, refPosMiss=0, refObsMiss=0, relPosHeadingValid=1, relPosNormalized=0)>

reported relPosLength = 8854041.078
reported relPosHeading = 271.7345
"""

from pyubx2 import haversine, bearing

BASE1 = {"name": "JoeSeelsGPS", "lat": 53.53, "lon": -1.33}
BASE2 = {"name": "ER_Pollybell_1", "lat": 53.43, "lon": -0.91}
ROVER = {"name": "pyubx2", "lat": 53.4506, "lon": -2.2403}

for base in (BASE1, BASE2):
    hdg = bearing(base["lat"], base["lon"], ROVER["lat"], ROVER["lon"])
    len = haversine(base["lat"], base["lon"], ROVER["lat"], ROVER["lon"]) * 100000
    print(
        f"name: {base['name']} ; estimated relPosLength: {len:.3f} ; estimated relPosHeading: {hdg:.4f}"
    )

```

```
name: JoeSeelsGPS ; estimated relPosLength: 6093366.427 ; estimated relPosHeading: 262.0256
name: ER_Pollybell_1 ; estimated relPosLength: 8823881.711 ; estimated relPosHeading: 272.0234
```

The estimates won't be exact because the base stations' positions are only given to 2 decimal places, but you can see that the NAV-RELPOSNED message is correctly reporting the relative heading and distance (+/- 0.6%) in each case.
